### ISSUE TITLE
Lazily create the code object when first viewed or executed

### ISSUE THREAD

**Issue by markshannon** [faster-cpython/ideas#464]
If well designed, the marshaled form of a code object should be compact. It is also cheap to load, with the potential for sharing across processes if memory mapped, although that is a bit trickier.

Many functions are never run. So why not keep the marshaled form of a code object in memory until we need the full form?
We can add a LazyCodeObject that looks like this:
```
struct PyCodeObject {
   HEADER
   const char *marshal_data;
   char padding[PAD_TO_SIZEOF_CODEOBJECT];
   _Py_CODEUNIT code[1]; // Set to UNMARSHAL_CODE_OBJECT instruction.
}
```
`PyFunction_GetCode()` would need to check for `LazyCodeObject` and unmarshal it if necessary.
The interpreter wouldn't need to do anything special as the `UNMARSHAL_CODE_OBJECT` instruction would unmarshal the code object.

Currently `PAD_TO_SIZEOF_CODEOBJECT` is rather large, but see https://github.com/faster-cpython/ideas/issues/463

**Comment by ericsnowcurrently** [comment]
This reminds me of work that @nascheme has done on lazy imports.

**Comment by gvanrossum** [comment]
IIRC I looked into this over a year ago and concluded that it wasn't feasible -- any changes to the marshal format proposed at the time were too sweeping or didn't have certain required features. What has changed? (I hope something has changed -- a good iterative design strategy is to remove the first bottleneck, then the next, etc.)

**Comment by markshannon** [comment]
"I concluded that it wasn't feasible" doesn't help much.
Why was it not feasible? What changes to the marshal format did it require?

**Comment by gvanrossum** [comment]
See https://github.com/faster-cpython/ideas/issues/66

**Comment by markshannon** [comment]
#66 seems a lot more complex.
All I'm really suggesting is deferring the unmarshal step.
The only complexity seems to be references. Did you perform any experiments to see what the effect of dropping references would be?

**Comment by gvanrossum** [comment]
A simpler version is https://github.com/faster-cpython/ideas/issues/64. Likely neither of these experiments worked out because I couldn't change the code object and the marshal format at the same time -- there would just be too much to change. A problem was that the interpreter bootstrap relies on the marshal and code formats.

If you want the long version of the story I would need to do some serious spelunking in the old branch where I did most of this work, https://github.com/gvanrossum/cpython/tree/pyco.

To be honest I'm not sure what you mean by "dropping references". Is this about `FLAG_REF` in marshal.c, or references to code objects, or references to other constants?

**Comment by markshannon** [comment]
By "dropping references" I mean dropping the ability of the marshaled data to refer to a previously unmarshaled object.
Doing so would mean that the unmarshaled graph was always a tree, never a DAG.
The loss of sharing could be a problem. Hopefully global sharing of immortal objects would mitigate that.

**Comment by gvanrossum** [comment]
Yeah, that's `FLAG_REF`, `TYPE_REF` and friends. IIRC I had to turn this off at some point to get an early version to work, and things kept working, but the marshal files bloated -- I don't recall by how much. The main source of sharing is various constants and tuples of constants and names, you'd be surprised how many tiny methods use the same set of names (e.g. def add(x, y), def sub(x, y), etc.).

The bloat means that loading *everything* would also be more expensive, and it could be more than the gain by compressing the bytecode. We should look into how much of the marshalled data is actually bytecode as opposed to other things (names, constants, line tables, etc.). But if you only load a subset the bloat might not be a problem.

Sharing pre-loaded global immortal objects would be possible but we'd have to have a way to reference them -- currently we only have `TYPE_NONE`, `TYPE_FALSE`, `TYPE_TRUE`, `TYPE_STOPITER` and TYPE_ELLIPSIS`.

Another problem was that you either have to read (portions of) the PYC file multiple times, or you have to keep the entire (raw, unmarshalled) contents of all/most PYC files in memory. A PYC file uses roughly the same amount of memory as the unmarshalled data, IIRC.

**Comment by markshannon** [comment]
I don't think this is ever going to happen, and lazy imports mean it is less needed.
